File: packages/i18nify-go/modules/masking/mask_phone_number_test.go
package masking

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestMaskPhoneNumber(t *testing.T) {
	tests := []struct {
		name    string
		phone   string
		opts    *MaskPhoneOptions
		want    string
		wantErr bool
	}{
		// ── default options (nil) ──────────────────────────────────────────
		{
			name:  "masks all but last 4 digits with default X",
			phone: "9876543210",
			opts:  nil,
			want:  "XXXXXX3210",
		},
		{
			name:  "preserves leading + with default options",
			phone: "+919876543210",
			opts:  nil,
			// digits: 919876543210 (12) → maskFrom=8 → XXXXXXXX3210
			want: "+XXXXXXXX3210",
		},
		{
			name:  "preserves formatting characters in their original positions",
			phone: "+1 234-567-8901",
			opts:  nil,
			// digits: 1,2,3,4,5,6,7,8,9,0,1 (11) → maskFrom=7 → X XXX-XXX-8901
			want: "+X XXX-XXX-8901",
		},

		// ── custom maskChar ────────────────────────────────────────────────
		{
			name:  "uses custom mask character",
			phone: "9876543210",
			opts:  &MaskPhoneOptions{MaskChar: "*"},
			want:  "******3210",
		},

		// ── custom visibleCount ────────────────────────────────────────────
		{
			name:  "shows last 6 digits when visibleCount is 6",
			phone: "9876543210",
			opts:  &MaskPhoneOptions{VisibleCount: 6},
			want:  "XXXX543210",
		},
		{
			name:  "masks nothing when visibleCount exceeds digit count",
			phone: "1234",
			opts:  &MaskPhoneOptions{VisibleCount: 6},
			want:  "1234",
		},
		{
			name:  "masks all digits when visibleCount is 0 (falls back to default 4)",
			phone: "9876543210",
			opts:  &MaskPhoneOptions{VisibleCount: 0},
			want:  "XXXXXX3210",
		},

		// ── combined custom options ────────────────────────────────────────
		{
			name:  "custom char and visibleCount together",
			phone: "+12025550199",
			opts:  &MaskPhoneOptions{MaskChar: "#", VisibleCount: 3},
			// digits: 12025550199 (11) → maskFrom=8 → ########199
			want: "+########199",
		},

		// ── error cases ───────────────────────────────────────────────────
		{
			name:    "returns error for empty string",
			phone:   "",
			opts:    nil,
			wantErr: true,
		},
		{
			name:    "returns error for whitespace-only string",
			phone:   "   ",
			opts:    nil,
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := MaskPhoneNumber(tt.phone, tt.opts)
			if tt.wantErr {
				assert.Error(t, err)
				return
			}
			assert.NoError(t, err)
			assert.Equal(t, tt.want, got)
		})
	}
}

File: packages/i18nify-js/src/modules/phoneNumber/__tests__/maskPhoneNumber.test.ts
import { maskPhoneNumber } from '../index';

describe('maskPhoneNumber', () => {
  describe('default masking (last 4 visible)', () => {
    it('masks a US number with separators', () => {
      expect(maskPhoneNumber('+1-800-555-1234')).toBe('+X-XXX-XXX-1234');
    });

    it('masks an E.164 Indian number', () => {
      expect(maskPhoneNumber('+919876543210')).toBe('+XXXXXXXX3210');
    });

    it('masks a plain 10-digit number (no country code)', () => {
      expect(maskPhoneNumber('9876543210')).toBe('XXXXXX3210');
    });

    it('preserves non-digit separators (spaces, hyphens, parens)', () => {
      expect(maskPhoneNumber('+44 (020) 1234-5678')).toBe(
        '+XX (XXX) XXXX-5678',
      );
    });
  });

  describe('options', () => {
    it('respects custom maskChar', () => {
      expect(maskPhoneNumber('+919876543210', { maskChar: '*' })).toBe(
        '+********3210',
      );
    });

    it('respects custom visibleCount = 6', () => {
      expect(maskPhoneNumber('+919876543210', { visibleCount: 6 })).toBe(
        '+XXXXXX543210',
      );
    });

    it('shows all digits when visibleCount >= total digit count', () => {
      expect(maskPhoneNumber('12345', { visibleCount: 10 })).toBe('12345');
    });
  });

  describe('edge cases', () => {
    it('handles number without leading +', () => {
      expect(maskPhoneNumber('18005551234')).toBe('XXXXXXX1234');
    });

    it('handles number with dot separators', () => {
      expect(maskPhoneNumber('+1.800.555.1234')).toBe('+X.XXX.XXX.1234');
    });
  });

  describe('input validation', () => {
    it('throws on empty string', () => {
      expect(() => maskPhoneNumber('')).toThrow();
    });
  });
});
